draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
+  ~~~
+
+  For clarity, it is generally preferred to use this formula directly, and
+  annotate its usage with this definition.
+
+- sqrt(x, q): The sqrt operation is a multi-valued function, i.e. there exist
+  two roots of x whenever x is a quadratic residue. To maintain compatibility
+  across implementations, a single-valued sqrt function is necessary. One way
+  to get such a function is by distinguishing a principal square root through a
+  predicate that only one of the roots holds.
+  Alternatively, an implementation of sqrt can use fixed formulas for its
+  calculation. For instance, in prime fields, the square root of a quadratic
+  residue x can be obtained as follows
+  - If q=3 (mod 4), sqrt(x, q) := x^((q+1)/4).
+  - If q=5 (mod 8), set z := x^((q+3)/8) and verify that z^2 = -x, if so, update
+    z = z\*sqrt(-1). Finally, sqrt(x, q) := z.
+
+  For extension fields, there exist methods that can be used in replacement,
+  see {{Adj13}}, {{SC85}}. Regardless the method chosen, the sqrt function
+  should be performed in constant time.
+
+- CMOV(a, b, c): If c=0, CMOV returns a, otherwise returns b. To prevent against
+  timing attacks, this operation must run in constant time without revealing the
+  value of c. Commonly some implementations assume the selector be c=1 or c=0;
+  thus, given a bitstring C, the desired selector c can be computed by OR-ing
+  all bits of C together. The resulting selector will be either 0 if all bits
+  of C are zero, or 1 if at least one bit of C is 1.
+
+- I2OSP and OS2IP: These functions are used to convert an octet string to
+  and from a non-negative integer {{RFC8017}}.
+
+- a \|\| b: denotes the concatenation of bitstrings a and b.
+
+# Hashing to a Finite Field {#hashtobase}
+
+The hash2base(x) function maps an arbitrary-large bitstring x into an element
+of a field F. Hence, it is parametrized by F and H, where H is a
+cryptographic hash function which outputs at least floor(log2(p)) + 1 bits.
+
+When q=p, the function first hashes x, converts the result to an integer, and
+reduces modulo p to produce a prime field element.
+When q=p^m, an element of a finite field is constructed using m prime field
+elements assuming that elements are in a polynomial representation.
+
+Uniformity: most algorithms assume that hash2base maps its input to the finite
+field uniformly. In practice, there will be inherent biases. For example,
+let H=SHA256 and F be a field of characteristic p = 2^255 - 19, then by reducing
+from a 256-bit output, the values from 2^256 - 38 to 2^256 - 1 have more
+probability to be chosen. In this example, the resulting bias is negligible,
+but for others this bias could be significant. This is a standard problem in
+generating uniformly distributed integers from a bitstring. In order to smooth
+out this bias, the hash2base algorithm should use a hash function that produces
+as many bits as possible before reducing modulo p. This approach is preferable
```

Comment:
nit: based on list feedback, "as many bits as possible" is not a meaningful requirement. We should probably be more precise here.

Reply:
Tracked on #99 .